### ISSUE TITLE
Data collection and storage on NAS for initial AI training.

### ISSUE THREAD

**Issue by michalharakal** [michalharakal/pAIpesense#4]
### Key Points
- A data flow diagram for the pAIpesense project can be created to illustrate how data is collected and processed during both training and production phases.
- In the training phase, data from controlled environments is collected, stored, and used to train the initial AI model.
- In the production phase, real-world data is processed on-device for leak detection, with user feedback enabling continuous model improvement.
- The diagram integrates seamlessly with the existing architecture, ensuring clarity for both developers and users.

### Overview
The pAIpesense project, designed for early water leak detection using IoT sensors and embedded AI, requires a clear data flow to support both initial model training and real-time operation in production. The data flow diagram provided below shows how audio and temperature data are collected, processed, and used to train and update the AI model, ensuring robust leak detection and continuous learning.

### Data Flow for Training
During the training phase, sensors collect data in controlled environments (e.g., lab setups with simulated leaks). This data is stored on a local Network Attached Storage (NAS) system and used to train the initial AI model, which is then deployed to the embedded AI device.

### Data Flow in Production
In production, sensors collect real-world data, which the embedded AI device processes using the trained model to detect leaks. If a leak is detected, alerts are sent to a web/mobile app. Users provide feedback (e.g., confirming or correcting detections), which is used to update the model on-device, supporting continuous learning.

### Integration with Architecture
The data flow aligns with the pAIpesense architecture, where sensors feed data to the embedded AI device, which interacts with storage, alerting systems, and smart home platforms. The diagram clarifies how data moves through these components during both phases.

---

# pAIpesense Data Flow Diagram

The pAIpesense project ([GitHub Repository](https://github.com/michalharakal/pAIpesense)) is an AI-based IoT device for early water leak detection, utilizing audio and temperature sensors with embedded AI for local processing. This document presents a detailed data flow diagram that illustrates how data is collected, processed, and utilized during both the training and production phases, supporting initial model development and continuous on-device learning.

## Data Flow Overview
The data flow for pAIpesense encompasses two primary phases:
- **Training Phase**: Data is collected in controlled environments to train the initial AI model, which is then deployed to the embedded AI device.
- **Production Phase**: Real-world data is collected and processed for real-time leak detection, with user feedback used to update the model on-device, enabling continuous learning.

The diagram below, created using Mermaid, visualizes the flow of data through the system, highlighting the distinct paths for training and production while showing their integration.

```mermaid
flowchart TD
    Sensors["Sensors<br>(Audio, Temperature)"] -->|Training Data| NAS[Training Data Storage]
    NAS --> ModelTraining[Model Training]
    ModelTraining --> InitialModel[Initial Model]
    InitialModel --> EmbeddedAI[Embedded AI Device]

    Sensors -->|Production Data| EmbeddedAI
    EmbeddedAI -->|Inference| Decision{Detect Leak?}
    Decision -->|Yes| Alerts[Local Alert & Notification]
    Decision -->|No| NoAction[No Action]
    Alerts --> WebApp[Web/Mobile App]
    WebApp -->|User Feedback| Feedback[Feedback]
    Feedback --> EmbeddedAI
    EmbeddedAI -->|Model Update| UpdatedModel[Updated Model]
```

### Components and Their Roles
The following table outlines the key components involved in the data flow, their roles, and how they function in training and production phases.

| **Component**                | **Role**                                                                 | **Training Phase**                                                                 | **Production Phase**                                                                 |
|------------------------------|--------------------------------------------------------------------------|-----------------------------------------------------------------------------------|-------------------------------------------------------------------------------------|
| **Sensors (Audio, Temperature)** | Collect raw audio (e.g., dripping, hissing) and temperature data from pipes. | Collect data in controlled environments (e.g., lab with simulated leaks).          | Collect data in real-world environments (e.g., homes, buildings).                    |
| **Training Data Storage (NAS)** | Stores raw and labeled data for model training and updates.               | Stores data from controlled environments for initial model training.               | Stores feedback-labeled data for potential future analysis or retraining.            |
| **Model Training**           | Develops the initial AI model using training data.                        | Uses data from NAS to train the model (likely on a separate system).               | Not active in production; updates occur on-device.                                   |
| **Initial Model**            | The trained AI model deployed to the device.                              | Deployed to the Embedded AI Device after training.                                 | Used as the starting point for inference in production.                              |
| **Embedded AI Device**       | Processes data, performs inference, and updates the model.                | Receives the initial model for deployment.                                         | Processes real-world data, detects leaks, and updates the model using feedback.      |
| **Inference**                | Analyzes sensor data to detect leaks.                                     | Not applicable in training.                                                       | Uses the current model to classify data as "leak" or "no leak."                     |
| **Local Alert & Notification** | Triggers alerts when a leak is detected.                                  | Not applicable in training.                                                       | Sends notifications to the Web/Mobile App for user awareness.                       |
| **Web/Mobile App**           | User interface for viewing alerts and providing feedback.                 | Not applicable in training.                                                       | Displays alerts and collects user feedback (e.g., confirming or correcting leaks).   |
| **Feedback**                 | User-provided data to label detections and improve the model.             | Not applicable in training (experts label data manually).                          | Provides labeled data (e.g., true/false leak confirmations) for model updates.       |
| **Updated Model**            | The AI model after incorporating feedback for continuous learning.        | Not applicable in training.                                                       | Deployed on the Embedded AI Device for improved inference.                           |

### Training Phase Data Flow
1. **Data Collection**:
   - Sensors (audio microphones and temperature sensors) collect data in controlled environments, such as a lab setup with various pipe materials (e.g., PVC, metal) and simulated leak scenarios (e.g., drips, cracks).
   - Audio data includes 30-second clips capturing leak sounds and background noises, sampled at a minimum of 8 kHz ([ScienceDirect](https://www.sciencedirect.com/science/article/pii/S0048969723001234)).
   - Temperature data is logged every 1-5 minutes to capture baseline conditions and anomalies.

2. **Data Storage**:
   - Raw data is stored on a local NAS system, organized into labeled datasets (e.g., "leak" vs. "no leak").
   - Metadata, such as timestamps and environmental conditions, is included for context.

3. **Model Training**:
   - Data from the NAS is used to train the initial AI model, likely on a separate system with sufficient computational resources.
   - The model is trained to recognize patterns in audio (e.g., dripping sounds) and temperature data (e.g., anomalies indicating leaks).

4. **Model Deployment**:
   - The trained model is deployed to the Embedded AI Device, ready for use in production.

### Production Phase Data Flow
1. **Data Collection**:
   - Sensors collect real-world data from pipes in homes or buildings, capturing audio and temperature variations in diverse environments.
   - Data collection may be triggered (e.g., audio recorded in bursts when temperature anomalies are detected) to conserve power.

2. **Inference**:
   - The Embedded AI Device processes sensor data using the current model (initial or updated) to detect potential leaks.
   - Inference results in a decision: "leak" or "no leak."

3. **Alerting**:
   - If a leak is detected, the Local Alert & Notification system sends alerts to the Web/Mobile App.
   - If no leak is detected, no action is taken.

4. **User Feedback**:
   - Users interact with the Web/Mobile App to view alerts and provide feedback (e.g., confirming a leak or marking a false positive).
   - Feedback is sent back to the Embedded AI Device as labeled data.

5. **Model Update**:
   - The Embedded AI Device uses feedback to update its model, employing techniques like incremental learning or online learning suitable for embedded systems (e.g., tinyML).
   - Updated data may also be stored on the NAS for logging or future retraining, though the project emphasizes on-device updates for continuous learning.

### Integration with Architecture
The data flow integrates with the pAIpesense architecture as follows:
- **Sensors**: Feed data to the Embedded AI Device, as shown in the architecture diagram ([GitHub Repository](https://github.com/michalharakal/pAIpesense)).
- **NAS**: Stores training and production data, aligning with the architecture’s Local NAS Storage component.
- **Embedded AI Device**: Central to both inference and model updates, consistent with its role in the architecture.
- **Alerts and Web/Mobile App**: Match the architecture’s Local Alert & Notification and Web/Mobile App components.
- **Smart Home Integration**: While not shown in the data flow diagram, the Embedded AI Device integrates with Home Assistant and the LMOS protocol, as described in the architecture.

### Best Practices
- **Data Diversity**: Collect data from various pipe materials, leak types, and environmental conditions to ensure model robustness.
- **Privacy**: Process audio data locally on the device to avoid storing or transmitting sensitive information, addressing privacy concerns.
- **Synchronization**: Timestamp audio and temperature data for combined analysis.
- **Continuous Learning**: Use user feedback to label new data, enabling on-device model updates without cloud dependency.
- **Data Augmentation**: Apply techniques like adding synthetic background noises to audio or simulating temperature fluctuations to enhance dataset diversity.

### Challenges and Considerations
- **Resource Constraints**: Embedded AI devices have limited computational power, so model updates must use lightweight algorithms (e.g., online gradient descent).
- **Data Quality**: Real-world data may be noisier than controlled data, requiring robust preprocessing and feature extraction.
- **Feedback Accuracy**: User feedback must be reliable to avoid introducing errors into the model.
- **Storage Management**: The NAS must handle large volumes of audio and temperature data efficiently, with clear organization for training and production datasets.

### Supporting Resources
- **GitHub Repository**: [pAIpesense](https://github.com/michalharakal/pAIpesense)
- **External Datasets**: Public datasets for leak detection, such as those with hydrophone audio, can supplement training data ([ScienceDirect](https://www.sciencedirect.com/science/article/pii/S0048969723001234)).
- **LMOS Documentation**: Supports smart home integration ([Eclipse LMOS](https://projects.eclipse.org/projects/iot.lmos)).

This data flow diagram and description provide a comprehensive view of how data is managed in the pAIpesense project, ensuring clarity for developers and stakeholders while supporting both initial training and continuous improvement in production.
